File: typescript/test/base64-crypto.test.ts
import { expect } from "chai";
import {
  encode,
  decode,
  urlSafeEncode,
  urlSafeDecode,
} from "../src/utils/crypto/base64";

describe("Crypto Base64 - Unit Test", () => {
  it("should encode and decode a string", () => {
    const input = "hello world";
    const encoded = encode(input);
    const decoded = decode(encoded);
    expect(decoded).to.equal(input);
  });

  it("should encode and decode a number", () => {
    const input = 12345;
    const encoded = encode(input);
    const decoded = decode(encoded);
    expect(decoded).to.equal(input);
  });

  it("should encode and decode an object", () => {
    const input = { foo: "bar", count: 42 };
    const encoded = encode(input);
    const decoded = decode(encoded);
    expect(decoded).to.deep.equal(input);
  });

  it("should encode and decode an array", () => {
    const input = ["a", "b", "c"];
    const encoded = encode(input);
    const decoded = decode(encoded);
    expect(decoded).to.deep.equal(input);
  });

  it("should decode object with items[] property as array", () => {
    const tupleLike = { items: [1, 2, 3] };
    const encoded = encode(tupleLike);
    const decoded = decode(encoded);
    expect(decoded).to.deep.equal({ items: [1, 2, 3] });
  });

  it("should urlSafeEncode and urlSafeDecode properly", () => {
    const input = { user: "alice", id: 99 };
    const encoded = urlSafeEncode(input);

    // base64url không có ký tự +, / hoặc = trong output
    expect(encoded).to.not.match(/[+/=]/);

    const decoded = urlSafeDecode(encoded);
    expect(decoded).to.deep.equal(input);
  });

  it("should urlSafeDecode object with items[] property as array", () => {
    const tupleLike = { items: ["x", "y"] };
    const encoded = urlSafeEncode(tupleLike);
    const decoded = urlSafeDecode(encoded);
    expect(decoded).to.deep.equal({ items: ["x", "y"] });
  });
});

File: python/test/base64_crypto.test.py
import os
import sys

BASE_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
sys.path.insert(0, os.path.join(BASE_DIR, "src"))

import unittest
import re
from utils.crypto.base64 import (
    encode,
    decode,
    url_safe_encode,
    url_safe_decode,
)


class TestCryptoBase64(unittest.TestCase):

    def test_encode_decode_string(self):
        input_data = "hello world"
        encoded = encode(input_data)
        decoded = decode(encoded)
        self.assertEqual(decoded, input_data)

    def test_encode_decode_number(self):
        input_data = 12345
        encoded = encode(input_data)
        decoded = decode(encoded)
        self.assertEqual(decoded, input_data)

    def test_encode_decode_object(self):
        input_data = {"foo": "bar", "count": 42}
        encoded = encode(input_data)
        decoded = decode(encoded)
        self.assertEqual(decoded, input_data)

    def test_encode_decode_array(self):
        input_data = ["a", "b", "c"]
        encoded = encode(input_data)
        decoded = decode(encoded)
        self.assertEqual(decoded, input_data)

    def test_decode_items_property_as_array(self):
        input_data = (1, 2, 3)
        encoded = encode(input_data)
        decoded = decode(encoded)
        self.assertEqual(decoded, (1, 2, 3))

    def test_url_safe_encode_decode(self):
        input_data = {"user": "alice", "id": 99}
        encoded = url_safe_encode(input_data)

        # base64url không có ký tự +, / hoặc =
        self.assertIsNone(re.search(r"[+/=]", encoded))

        decoded = url_safe_decode(encoded)
        self.assertEqual(decoded, input_data)

    def test_url_safe_decode_items_property_as_array(self):
        input_data = ("x", "y")
        encoded = url_safe_encode(input_data)
        decoded = url_safe_decode(encoded)
        self.assertEqual(decoded, ("x", "y"))


if __name__ == "__main__":
    unittest.main()
